functional/stateful-types.md

```diff
@@ -0,0 +1,269 @@
+# Stateful Types: Generics as Type Classes
+
+## Description
+
+Rust's functional roots allow for it to be more expressive in the type system
+than many other languages, and turn many kinds of programming problems into
+"static typing" problems. A key part of this idea is the way generic types work.
+
+In C++ and Java, for example, generic types are a meta-programming construct for
+the compiler. `Vec<int>` and `Vec<char>` in C++ are just two different copies of
+the same boilerplate code for a `Vec` type, with two different types filled in.
+
+In Rust, the generic type parameter creates what is known as a "type class
+constraint", and each value used by an end user *actually changes the type of
+each instatiation*. In other words, `Vec<usize>` and `Vec<char>` *are two
+different types*.
+
+This is why `impl` blocks must specify generic parameters: different ones can
+have different `impl` blocks on them.
+
+It is recommended in rust to use generic types to enforce invariants. The best
+example is a state machine.
+
+## Example
+
+Suppose you are designing an interpreted language runtime, which requires a JIT
+compiler in order to process. Many Rust novices coming from other languages
+would implement it this way:
+
+```rust
+#[derive(Debug)]
+pub enum CompileError {
+    // error type, std::error::Error impl skipped for brevity
+}
+
+#[derive(Debug)]
+pub enum ExecError {
+    // error type, std::error::Error impl skipped for brevity
+}
+
+pub type ExecutionResult = Result<(), ExecError>;
+
+pub type CompileResult = Result<(), CompileError>;
+
+pub enum Param {
+    GCSize(usize),
+    MaxStack(usize),
+    // more parameters...
+}
+
+#[derive(Default)]
+pub struct Interpreter {
+    // state and execution data
+}
+
+impl Interpreter {
+    /// Set a system prameter.
+    ///
+    /// # Panics
+    /// Will panic if called after a script is executed.
+    ///
+    pub fn set_param(&mut self, p: Param) {
+        /* update the state based on the parameter */
+    }
+
+    /// Set up a new execution environment.
+    ///
+    /// # Panics
+    /// Will panic if called more than once or after a script has been compiled.
+    ///
+    pub fn init(&mut self) {
+        /* Create the heap, prepare for use... */
+    }
+
+    /// Load and compile a script. Returns any errors encountered.
+    ///
+    pub fn compile_script(&mut self, script: &str) -> CompileResult {
+        /* actually do the compile... */
+        Ok(())
+    }
+
+    /// Execute all scripts added. Returns any errors encountered.
+    ///
+    /// # Panics
+    /// Will panic if zero scripts have been compiled.
+    ///
+    pub fn exec(&mut self) -> ExecutionResult {
+        /* actually run the script... */
+        Ok(())
+    }
+}
+
+fn main() {
+    let mut interp = Interpreter::default();
+    interp.set_param(Param::GCSize(1024 * 1024 * 1024));
+    interp.init();
+    interp.compile_script("print('2 + 2')").unwrap();
+    interp.exec().unwrap();
+}
+```
```

Comment:
Example seemed complicated. If someone like me who had no experience with state machine, this will be hard to understand. Just wondering could there be a simpler example?

Say, empty letter -> written letter -> closed letter?